### PR TITLE
Torch

### PR DESCRIPTION
**Pull request by akoryagin**
This pr adds `torch`-based  model and `solver` and removes `tensorflow` completely.
Leaves only one tutorial covering wide range of examples; updates `pylint` to working state.

Classes added by this pr:
* `TorchModel`
* `Solver`
* `ConvBlockModel` - model based on `conv_block` from `batchflow`.

**NOTE**: despite the diff looking complex, there are basically two files in the library: `model_torch.py` and the tutorial notebook.

### CODE REVIEW

File: pydens/model_torch.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+            # Sample batch of points and compute solution-approximation on the batch.
+            if sampler is None:
+                xs = [torch.rand((batch_size, 1)) for _ in range(self.model.total)]
+            else:
+                xs_concat = sampler.sample(batch_size).astype(np.float32)
+                xs = [torch.from_numpy(xs_concat[:, i:i+1]) for i in range(xs_concat.shape[1])]
+            for x in xs:
+                x.requires_grad_()
+            u_hat = self.ctx.run(self.model, *xs)
```
Comment (SergeyTsimfer) [inline pydens/model_torch.py:190]: Making one `(B, N)-shaped` tensor would make your life a lot easier: no weird starred expressions all over the place

File: pydens/model_torch.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+            for num in nums_constraints:
+                loss += criterion(self.ctx.run(self.constraints[num], self.model, *xs),
+                                  torch.Tensor([0.0]))
+
+            # Optimizer step.
+            loss.backward()
+            self.optimizer.step()
+
+            # Gather and store training stats.
+            self.losses.append(loss.detach().numpy())
```
Comment (SergeyTsimfer) [inline pydens/model_torch.py:208]: `.cpu()`?

File: pydens/model_torch.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        Points are given by a list of tensors.
+        """
+        xs = list(xs)
+        for i, x in enumerate(xs):
+            if not isinstance(x, torch.Tensor):
+                xs[i] = torch.Tensor(np.array(x)).float()
+
+        self.model.eval()
+        result = self.ctx.run(self.model, *xs)
+        return result.cpu().detach().numpy()
```
Comment (SergeyTsimfer) [inline pydens/model_torch.py:221]: `detach` -> `cpu` -> `numpy`

File: pydens/model_torch.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+                                  torch.Tensor([0.0]))
+
+            # Optimizer step.
+            loss.backward()
+            self.optimizer.step()
+
+            # Gather and store training stats.
+            self.losses.append(loss.detach().numpy())
+
+    def solve(self, *xs):
```
Comment (SergeyTsimfer) [inline pydens/model_torch.py:210]: I am definitely voting for the `predict` name

File: pydens/model_torch.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+            # Apply transformation to bind the initial condition.
+            if self.initial_condition is not None:
+                _xs_spatial = [xs_spatial[:, i] for i in range(xs_spatial.shape[1])]
+                u = ((nn.Sigmoid()(t / torch.exp(self.log_scale)) - .5) * u
+                     + self.initial_condition(*_xs_spatial).view(-1, 1))
+            return u
+        return func
+
+class ConvBlockModel(TorchModel):
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply pydens/model_torch.py:101]: As I finished the entire `model_torch.py` file, I actually don't even think that you need one model class, nevermind two of them: all you need is to store problem-related parameters in the `Solver`, make `anzatz` a method (as it is a deterministic function of its parameters), and have `ConvBlock` as an attribute of the `Solver`-instance

File: pydens/model_torch.py (hunk 1 above)
Comment (akoryagin) [reply pydens/model_torch.py]: Fix it in most places - when applying `model`.

Still needed to run `equation` and `constraint` so left it there.

File: pydens/model_torch.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        for layer in layers:
+            for param in getattr(self, layer).parameters():
+                param.requires_grad = True
+
+        # Unfreeze variables.
+        for variable in variables:
+            param = getattr(self, variable)
+            param.requires_grad = True
+
+    def forward(self, *xs):
```
[the remark this answers is not in the document]
Comment (akoryagin) [reply pydens/model_torch.py]: Remove it from  base `TorchModel`

File: pydens/model_torch.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        """ Pass variables through the network and apply anzatc-transformation to bind
+        initial condition. """
+        xs = [x.view(-1, 1) for x in xs]
+        xs = torch.cat(xs, dim=1)
+        return xs
+
+    def anzatc(self):
+        """ Make transform of the model-output needed for binding initial and boundary conditions. """
+        def func(u, xs):
```
[the remark this answers is not in the document]
Comment (akoryagin) [reply pydens/model_torch.py:75]: Fix it

File: pydens/model_torch.py (hunk 2 above)
Comment (akoryagin) [reply pydens/model_torch.py:229]: Fix

File: pydens/model_torch.py (hunk 3 above)
Comment (akoryagin) [reply pydens/model_torch.py:241]: Fix
